src/api/gamestate.cc

```diff
@@ -48,6 +101,85 @@ mahjong::GameState* AdvanceGameState(mahjong::GameState* state) {
   return new_state.release();
 }
 
+CObservedGameState ObserveGameState(mahjong::GameState* state) {
+  CObservedGameState observed = {};
+  
+  if (!state) {
+    return observed;
+  }
+  
+  // Straightforward copies
+  observed.currentPlayer = state->currentPlayer;
+  observed.turnNum = state->turnNum;
+  observed.roundNum = state->roundNum;
+  observed.riichiSticks = state->riichiSticks;
+  observed.counters = state->counters;
+  observed.lastCall = state->lastCall;
+  observed.lastCaller = state->lastCaller;
+  observed.concealedKan = state->concealedKan;
+  observed.seed = state->seed;
+  observed.pendingPiece = static_cast<CPiece>(state->pendingPiece.toUint8_t());
+  
+  // State function enums
+  observed.prevState = ConvertStateFunctionType(state->prevState);
+  observed.currState = ConvertStateFunctionType(state->currState);
+  observed.nextState = ConvertStateFunctionType(state->nextState);
+  
+  // Player data
+  for (int i = 0; i < 4; i++) {
+    observed.scores[i] = state->scores[i];
+    observed.points[i] = state->players[i].points;
+    observed.hasRonned[i] = state->hasRonned[i];
+    
+    // Convert Hand to CHand
+    const auto& cpp_hand = state->hands[i];
+    CHand& c_hand = observed.hands[i];
+    
+    // Live pieces
+    c_hand.liveCount = static_cast<int>(cpp_hand.live.size());
+    for (int j = 0; j < kMaxLiveHandSize; j++) {
```

Comment:
I would just set the pieces for live count and leave the remaining array slots as garbage.